src/service/viewport-impl.js

```diff
@@ -546,33 +549,52 @@ export class Viewport {
   }
 
   /**
-   * @param {{paddingTop: number, duration: (number|undefined), curve: (string|undefined)}} event
+   * @param {!JSONType} event
    * @private
    */
   updateOnViewportEvent_(event) {
     this.binding_.updateViewerViewport(this.viewer_);
-    const paddingTop = event.paddingTop;
+    const paddingTop = event['paddingTop'];
+    const duration = event['duration'];
+    const curve = event['curve'];
+    const transient = event['transient'];
+
     if (paddingTop != this.paddingTop_) {
-      const lastPaddingTop = this.paddingTop_;
+      this.lastPaddingTop_ = this.paddingTop_;
       this.paddingTop_ = paddingTop;
-      this.binding_.updatePaddingTop(this.paddingTop_, /* adjustScroll */true,
-          lastPaddingTop);
-      this.fixedLayer_.updatePaddingTop(this.paddingTop_);
-
-      if (event.duration > 0) {
-        // Add transit effect on position fixed element
-        const tr = numeric(lastPaddingTop - this.paddingTop_, 0);
-        Animation.animate(this.ampdoc.getRootNode(), time => {
-          const p = tr(time);
-          this.fixedLayer_.transformMutate(`translateY(${p}px)`);
-        }, event.duration, event.curve).thenAlways(() => {
-          this.fixedLayer_.transformMutate(null);
+      if (this.paddingTop_ < this.lastPaddingTop_) {
+        this.binding_.hideViewerHeader(transient, this.lastPaddingTop_);
+        this.animateFixedElements_(duration, curve);
+      } else {
+        this.animateFixedElements_(duration, curve).then(() => {
+          this.binding_.showViewerHeader(transient, this.paddingTop_);
         });
       }
     }
   }
 
   /**
+   * @param {number} duration
+   * @param {string} curve
+   * @return {!Promise}
+   * @private
+   */
+  animateFixedElements_(duration, curve) {
+    this.fixedLayer_.updatePaddingTop(this.paddingTop_);
+    if (duration > 0) {
```

Comment:
Might be cleaner to exit early if `duration <= 0`, and unnest the conditional we currently have.